File: day-3/go/part1.go
package day3

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Day3 struct{}

type Coord struct {
	x int
	y int
}

func (d Day3) Part1(filename *string) string {
	start := time.Now()
	content, _ := os.ReadFile(*filename)
	lines := strings.Split(string(content), "\n")

	// initialize engine
	var total_y int = len(lines)
	var total_x int = len(lines[0])
	var engine [][]int = make([][]int, total_y)

	for i := 0; i < total_y; i++ {
		engine[i] = make([]int, total_x)
	}

	var special_chars []Coord

	// iterate through engine to find valid words
	for y, row := range lines {
		var num = 0
		var num_start = 0

		for x, char := range row {
			var parsed_num, _ = strconv.Atoi(string(char))
			var is_number bool = char >= '0' && char <= '9'

			if is_number {
				// add to current number
				num = num*10 + parsed_num
			} else {
				// if not number then add curren numbet
				// to engine at all positions starting from num_start
				if num > 0 {
					for i := num_start; i < x; i++ {
						engine[y][i] = num
					}
				}

				// reset current number and num_start
				num = 0
				num_start = x + 1
			}

			// if there is a special char in the surrounding
			if !is_number && char != '.' {
				special_chars = append(special_chars, Coord{x, y})
			}
		}

		// check for last number
		if num > 0 {
			for i := num_start; i < total_x; i++ {
				engine[y][i] = num
			}
		}
	}

	// iterate through special chars to find the total part numbers
	var total_part_numbers int = 1

	for _, coord := range special_chars {
		var directions = []Coord{
			{-1, 0}, {1, 0}, // down, up
			{0, -1}, {0, 1}, // right, left
			{-1, -1}, {1, 1}, // down-right, up-left
			{-1, 1}, {1, -1}, // down-left, up-right
		}

		// iterate through directions to find numbers that surround
		// the special char
		for _, direction := range directions {
			var new_y = coord.y + direction.y
			var new_x = coord.x + direction.x

			// make sure that y and x are within the bounds of the engine
			if new_y < 0 || new_y >= total_y || new_x < 0 || new_x >= total_x || engine[new_y][new_x] == 0 {
				continue
			}

			// if the number is found then add it to the total part numbers
			total_part_numbers += engine[new_y][new_x]

			// remove the number from the engine so that it won't be counted again
			// also remove the places where the number is found
			engine[new_y][new_x] = 0

			// remove the lef and right part of the new_x and new_y
			// so that it won't be counted again
			for x := new_x + 1; x < total_x && engine[new_y][x] > 0; x++ {
				engine[new_y][x] = 0
				break
			}

			for x := new_x - 1; x >= 0 && engine[new_y][x] > 0; x-- {
				engine[new_y][x] = 0
				break
			}

		}
	}

	t := time.Now()
	elapsed := t.Sub(start)

	fmt.Println("Part 1 took", elapsed)

	return fmt.Sprint(total_part_numbers)
}

File: day-3/cpp/part1.cpp
// To run do: c++ part1.cpp -Wall -std=c++17 -o output1.out && ./output1.out
#include <string>
#include <vector>
#include <fstream>
#include <iostream>
#include <map>
#include "../../utils/utils.cpp"

using namespace std;

// check for all special characters in the surrounding of a given position
bool special_char_in_surrounding(int row, int col, std::vector<std::vector<char>> grid)
{
    bool has_special_char = false;

    // Check if the surrounding characters are special characters
    for (int i = row - 1; i < row + 2; i++)
    {
        for (int j = col - 1; j < col + 2; j++)
        {
            // Check if the current position is out of bounds
            if (i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size())
            {
                continue;
            }
            char curr_char = grid[i][j];

            if (!isdigit(curr_char) && curr_char != '.')
            {
                has_special_char = true;
            }
        }
    }

    return has_special_char;
}

int main()
{
    // Initialize a vector to store the file input and set the input file name.
    std::string fileName = "../input.txt";
    std::vector<std::string> lines = read_file(fileName);

    // Initialize a 2d vector to store the engine
    std::vector<std::vector<char>> engine;
    int rows = lines.size();
    int cols = lines[0].size();

    // Initialize the engine
    for (int i = 0; i < rows; i++)
    {
        std::vector<char> row;
        for (int j = 0; j < cols; j++)
        {
            row.push_back(lines[i][j]);
        }
        engine.push_back(row);
    }

    int total_part_numbers = 0;

    // Iterate through the engine to find valid words
    for (int row_num = 0; row_num < rows; row_num++)
    {
        // Look for a number in the line, add the number to the total
        int curr_num = 0;
        bool is_valid_part = false;
        for (int col_num = 0; col_num < cols; col_num++)
        {
            char curr_char = engine[row_num][col_num];
            bool is_digit = isdigit(curr_char);
            // check if the current character is a number
            if (is_digit)
            {
                curr_num = curr_num * 10 + (curr_char - '0');
            }
            else
            {
                // If the character is not a number we reset the current number
                // and check if the current number is valid
                if (is_valid_part && curr_num > 0)
                {
                    total_part_numbers += curr_num;
                }

                curr_num = 0;
                is_valid_part = false;
            }

            if (is_digit && special_char_in_surrounding(row_num, col_num, engine))
            {
                is_valid_part = true;
            }
        }

        // check last number
        if (is_valid_part && curr_num > 0)
        {
            total_part_numbers += curr_num;
        }
    }

    // Print the result.
    std::cout << "Result: " << total_part_numbers << std::endl;
}
